\chapter{Bioinformatics}

\section{Computational problems in bioinformatics}

In this chapter, I present an overview of some interesting and important bioinformatical problems, particularly protein folding, which is one of the central problems related to drug discovery\cite{BockenhauerAlgoBioinfo}.

\subsection{DNA sequencing and the Human Genome Project}

The complete genetic material of a human being is called the human genome, which is contained in the nucleus of every single cell of the human body and is replicated during cell division. The human genome consists of 46 chromosomes, each containing a single deoxyribonucleic acid, or DNA molecule for short. A DNA molecule is two twisting, paired strands (called the double helix structure) and comprises around 3.2 billion pairs of nucleotide bases. Nucleotides are usually denoted by the letters A, T, C and G, with A-T and C-G forming pairs. A sequence of nucleotides that together encode the synthesis of a specific protein or RNA is called a gene.

Research into the human genome started with the discovery of the double helix structure of the DNA molecule in 1953 by Francis Crick and James Watson. The Human Genome Project\cite{CollinsHumanGenome1995} was launched in 1990, an international research effort to identify the function of all 50,000 to 100,000 genes of the complete human genetic material and was completed in early 2022\cite{zahn_filling_2022}. Besides injuries, nearly all human medical conditions are related to mutations at specific locations in the structure and function of the genetic material\cite{CollinsHumanGenome1995}, so identifying changes in an individual's DNA sequence can be an indispensable tool for predicting and preventing disease while providing individualized care. This requires a quick sequencing of specific parts of a person's DNA and comparing them to a healthy variant.

\begin{figure}[H]
    \centering
    \includegraphics[width=0.6\textwidth]{figures/bioinformatics/dna.png}
    \caption{DNA structure\cite{dna_sequencing}}
\end{figure}

From a computer science viewpoint, a simplified DNA model is a string consisting of the four letters from one of the strands of the physical molecule. (However, this model omits the 3D spatial structure.) Due to the limits of our technology, current methods enable us to read up to 1000 consecutive letters from a single sample. To sequence a longer string, we can break the molecule into sufficiently smaller fragments and sequence them individually. Unfortunately, in this process, the order in which the fragments originally occurred is lost. One possible approach to deal with this problem is to create copies of the DNA molecule we are interested in and randomly break each of these copies into fragments. With a high probability, the resulting fragments from different copies will overlap with each other. Then, the task becomes reconstructing the original string from these fragments while considering the possible errors in our biological measurements. The possible number of fragment orderings is enormously large, so a brute force algorithm does not suffice for solving this task.\cite{CollinsHumanGenome1995}

Between any two humans, the amount of genetic variation is estimated to be around 0.6\%\cite{the_1000_genomes_project_consortium_global_2015}. This enables us to use the targeted part of the genome sequence preassembled by the Human Genome Project and align our obtained DNA sequence fragments to the best fitting position to obtain the entire string of our DNA. Algorithms such as the Smith-Waterman algorithm and the Needleman-Wunsch algorithm can efficiently perform this sequence alignment task.

\subsection{Protein folding}

Proteins are one of the fundamental building blocks of the human body. They play an essential role in our immune system response, transportation of molecules throughout the body, the catalysation of metabolic reactions and signal transmission. Protein molecules consist of a single, long chain of amino acids. While several amino acid molecules exist in nature, only 20 of those are present in proteins.\cite{BockenhauerAlgoBioinfo}

The 20 standard amino acids can be seen in Table \ref{aminoacids20}. The most important property of these is their affinity to water: Some amino acids are polar (hydrophilic), commonly denoted by a P, which means they can establish hydrogen bonds with $H_2O$ molecules. In contrast, others are hydrophobic (nonpolar), commonly denoted by an H, which do not establish these bonds with water.

\begin{table}[H]
\centering
\begin{tabular}{|l|ccc|}
\hline
Name          & Abbreviation & Code & Polarity \\
\hline
Alanine       & Ala & A & H \\
Valine        & Val & V & H \\
Leucine       & Leu & L & H \\
Isoleucine    & Ile & I & H \\
Phenylalanine & Phe & F & H \\
Proline       & Pro & P & H \\
Methionine    & Met & M & H \\
Tryptophan    & Trp & W & H \\
\hline
Arginine      & Arg & R & P \\
Asparagine    & Asn & N & P \\
Aspartic acid & Asp & D & P \\
Cysteine      & Cys & C & P \\
Glutamic acid & Glu & E & P \\
Glutamine     & Gln & Q & P \\
Glycine       & Gly & G & P \\
Histidine     & His & H & P \\
Lysine        & Lys & K & P \\
Serine        & Ser & S & P \\
Threonine     & Thr & T & P \\
Tyrosine      & Tyr & Y & P \\
\hline
\end{tabular}
\caption{Aminoacids in proteins}
\label{aminoacids20}
\end{table}

The functionality and role of a protein are determined by its spatial structure, of which four levels are distinguished.

\textbf{Primary structure}: The sequence of amino acids in the protein.

\begin{figure}[H]
    \centering
    \includegraphics[width=\textwidth]{figures/bioinformatics/protein_structure_primary.png}
    \caption{Primary structure of the protein PCNA\cite{protein_structure}}
\end{figure}

\textbf{Secondary structure}: Created by the interactions between the atoms along the chain of amino acids, forming substructures ($\alpha$-helices, $\beta$-sheets, loops).

\begin{figure}[H]
    \centering
    \includegraphics[width=\textwidth]{figures/bioinformatics/protein_structure_secondary.png}
    \caption{Secondary structure of the protein PCNA\cite{protein_structure}}
\end{figure}

\textbf{Tertiary structure}: The spatial arrangement of all atoms within the chain. Secondary structure elements are grouped together as motifs, and functional units called domains.

\begin{figure}[H]
    \centering
    \includegraphics[width=\textwidth]{figures/bioinformatics/protein_structure_tertiary.png}
    \caption{Tertiary structure of the protein PCNA\cite{protein_structure}}
\end{figure}

\textbf{Quaternary structure}: Describes the composition of the whole protein from polypeptide subunits and potentially other molecular parts.

\begin{figure}[H]
    \centering
    \includegraphics[width=\textwidth]{figures/bioinformatics/protein_structure_quaternary.png}
    \caption{Quaternary structure of the protein PCNA\cite{protein_structure}}
\end{figure}

For example, hemoglobin is a protein found in red blood cells whose function is to carry oxygen molecules throughout the blood vessels. The oxygen molecule binds to the heme groups found in the protein.

\begin{figure}[H]
    \centering
    \includegraphics[width=0.5\textwidth]{figures/bioinformatics/hemoglobin.png}
    \caption{Hemoglobin, the iron-containing heme groups are shown in green\cite{wheeler_1gzx_nodate}}
\end{figure}

To understand the function of a protein, we must determine its complete 3D structure. While techniques exist to measure a molecule's structure in real life, these are incredibly time-consuming and expensive to perform, even for a single molecule. Consequently, we would like to design algorithms that can predict the complete 3D structure of a protein based on its primary structure or its amino acid sequence. These algorithms are called protein folding algorithms.

In the molecular docking task, the candidate protein has already been selected with its 3D structure known. For the purpose of drug discovery, we are searching for the best possible protein as well, which is why databases of proteins with their predicted 3D structures are being worked on, such as \href{https://www.uniprot.org/}{https://www.uniprot.org/}, or \href{https://alphafold.ebi.ac.uk/}{https://alphafold.ebi.ac.uk/} \cite{senior_improved_2020}.

Modelling and predicting the complete four levels of protein structure is a complex task. Interestingly, simplified models exist that can be experimentally shown to achieve a reasonable approximation. One of these models was given by Dill et al. in 1995. (republished in 2008) \cite{dill_principles_2008}. Even though these models vastly simplify the real-life behaviour of protein chains, the corresponding algorithmic problem is still NP-hard. Later in my dissertation, I will further discuss Dill's HP model for protein folding.

\subsection{Molecular docking}

Proteins are the primary agents of biological function, as they control the various chemical mechanisms that occur inside the cells. Proteins that act as biological catalysts are called enzymes. Catalysts facilitate various chemical reactions without being consumed in the process. These molecules have a binding site (a 'hole' on their 3D surface), into which only a specific other molecule fits, which is called a substrate. During the chemical reaction, the substrate is turned into other products.  \cite{fionda_networks_2019}

Many human diseases result from abnormal interactions between proteins. In order to prevent disease, we can stop (inhibit) these interactions from happening. This is done by blocking the binding site of the faulty enzyme with another molecule. Conventional medicine uses giant molecules (antibodies) or tiny molecules (like aspirin) to achieve this. The next generation of protein therapeutics currently under development aims to find inhibitors within the family of smaller-sized proteins.
 \cite{ryan_proteinprotein_2005}

In order to inhibit an enzyme's reaction from happening, we must find a protein that folds into a shape that fits into the enzyme's binding site to prevent it from catalyzing a reaction. The analogy of 'lock-and-key' was coined for this by Emil Fischer in 1894. He suggested that the 'lock' describes the enzyme's binding site, and the 'key' describes the missing molecule inhibitor, which has to fit into the 'lock'. \cite{a_molecular_2018} \cite{walker_molecular_2008} The computational task of predicting whether a protein with a known 3D structure will fit inside the binding site of a specific enzyme is called molecular docking.

\begin{figure}[H]
    \centering
    \includegraphics[width=0.8\textwidth]{figures/bioinformatics/molecular_docking.jpg}
    \caption{Molecular docking\cite{chaos_chemical_2008}}
\end{figure}

In the molecular docking problem, both the binding site's 3D structure and the protein's 3D structure are represented by a graph. The graph's vertices are the atoms on the surface. The edges represent a chemical bond between the corresponding atoms. The weight of a given edge represents the physical distance between the atoms. Omitting the edge weights, the problem is analogous to subgraph isomorphy (an NP-hard problem), the binding site's graph representation being a subgraph of the inhibitor molecule's entire surface in a graph representation. With the edge weights present, we can employ a measure that assigns a score to an isomorphy mapping of the vertices between the two graphs. Typically root mean square deviation (RMSD) is used, which can be defined for two sets of (ordered) points. The mapping achieving the highest score is the best theoretical fit for the molecule.
 \cite{wang_protein_2021}

\section{The connection between bioinformatics and quantum algorithms}

While some problems in bioinformatics, such as DNA sequence alignment, already have a fast enough solution, others, such as protein folding and molecular docking, turn out to be NP-hard ones. I have chosen to explore protein folding in more detail.

Grover's search algorithm is a quantum algorithm that can solve problems classically in NP if a candidate solution can be encoded to a quantum register and the verifier algorithm can be translated to quantum gate logic. Applying multiple Grover searches, we can also give a solution that minimizes or maximizes a target function. In the case of protein folding, a candidate solution is a possible orientation of a chain in 3D space, and we are looking for the minimum energy configuration.

Quantum walks work by exploring the problem's domain, which can be represented in a graph form. The vertices are candidate solutions, and the edges connecting them are small transformations on them. In the case of protein folding, a possible small transformation is rotating the chain at a single point. This operation also has to be translated to quantum gate logic.

These two algorithms can be effectively combined. For example, the initialization step for Grover can be a quantum walk, or they can be applied in turns.