\chapter{Conclusion}

In this dissertation I have covered three main topics: bioinformatics and protein folding, quantum walks, and Grover's search algorithm.

\section{Achievements}

My research has started with understanding quantum computing in general, focused specifically on quantum walks. I have read several excellent sources \cite{Aharonov, KempeIntroduction, Portugal, Santha, XiaReview} and lecture notes, and found \cite{Portugal} to be a comprehensive general introduction. I reformulated its descriptions to the language of matrices in an explicit manner, which matches both classical random walk descriptions and universal quantum computing hardware requirements. I believe these are easier to understand for someone with a college-level software engineering background who is new to the subject.

Besides this, I gave a generalized requirement for constructing quantum random walks on $d$-regular graphs employing the position-coin notation and improved the memory requirement of $n$-dimensional lattice walks. I presented my proofs for both of these advancements.

I have written a quantum walk simulator software, which helped me both understand quantum algorithms in general, as well as this particular algorithm. I have started looking into usecases for quantum walks and found their connection to Grover's search algorithm and solving NP-hard problems in bioinformatics.

I have looked into three bioinformatical problems: DNA sequencing, protein folding and molecular docking and explained their connection to computational problems. I am interested in computer-aided drug design, so I have chosen to further investigate protein folding and have explored Dill's simplified model for a protein structure: the HP model, which is still NP-hard, but can be implemented on a gated general-purpose quantum computer.

I have devised a way to encode a protein in this model in a quantum register, however I have ran into a problem with formulating a verifier algorithm. The simulator frameworks used too much memory, even on the much simpler, but analogous problem of Sudoku. I estimated, that the largest computable protein chain in Qiskit would contain at most four amino acids, on which the problem is trivial.

After further looking into why my incomplete solution in Qiskit uses such large amounts of memory, I have found many inefficiencies in the software, in terms of memory handling. This is due to the fact, that quantum frameworks, such as Qiskit (IBM) and Cirq (Google) focus on a different goal: to allow the programming of quantum computers their respective vendors sell, which means that simulation, especially memory-efficient simulation, is not their primary concern. Their implementation uses sparse-matrix representation of each unitary matrix and allocates the resources for each individual instance of them. This made it very difficult and expensive to use them for my usecase, which is testing my quantum algorithms for protein folding even on relatively small inputs.

This is why I have decided to write my own quantum simulator framework, that focuses on memory efficient simulation.

The primary goal of this framework is to reduce memory-usage of simulation while trading in runtime. For research purposes, it is acceptable to wait for example a few days for a simulation of protein folding runs on a relatively high-end PC, however it is not cost-effective to buy terabytes of memory or rent a memory-optimized virtual machine from the cloud.

In this dissertation, I have developed the mathematical framework for implementing general-purpose software for gated quantum computer simulations. These developments have been:

\begin{itemize}
    \item The logic of handling the probability amplitudes in the current set of registers and applying operators to a subset of these registers using qubit mapping permutations on their binary indexing sequences.
    \item The architecture allows individual tricks for memory-efficient operator implementation, such as on-the-fly generation and the $u$ function method.
    \item The building blocks for Grover's algorithm's implementation, the Hadamard, Grover, Sum and MCnot operators.
\end{itemize}

\section{Plans for the future}

I would like to further develop my framework and use it to implement quantum protein folding in it using Grover's search algorithm framework and quantum walking.

In addition, I would like to introduce unit testing for the individual components of the software. Since all of these operators rely heavily on custom implementation, it is important that their correctness is verified. In particular, I would like to explore methamorphic testing, in which the operators are tested by verifying if they admit to certain mathematical properties, such as the self-adjointness of the Hamilton-operator.

Furthermore, I would like to extend the available operators in the framework so that other types of algorithms can be implemented in it as well.
